### PR TITLE
[TRA-508] Require that market exists in market map when creating new oracle market

### PR DESCRIPTION
**Pull request by chenyaoy**
### Changelist
[Describe or list the changes made in this PR]

### Test Plan
[Describe how this PR was tested (if applicable)]

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

- **New Features**
	- Enhanced integration of market map functionality within the application, improving interaction with prices and assets modules.
	- Expanded configurations for cryptocurrency markets in the sample pregenesis file, allowing for better integration with trading platforms.
	- Improved simulation of market parameters with dynamically generated configurations.
	- Integrated new keeper for market map management and expanded testing capabilities.

- **Bug Fixes**
	- Adjusted initialization processes to ensure accurate market state management within tests.

- **Tests**
	- Expanded test cases to include additional checks, updated function signatures, and improved robustness in handling market data.
	- Enhanced testing framework by incorporating comprehensive market condition simulations.

- **Refactor**
	- Modified function signatures and internal logic across multiple modules to accommodate new dependencies and enhance maintainability.
	- Standardized command syntax in the Makefile for consistency with current Docker practices.
	- Updated tests for market creation to include checks against the market map, enhancing validation of expected behavior.

### CODE REVIEW

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 1]
```diff
@@ -298,11 +302,26 @@ func TestAddNewMarketProposal(t *testing.T) {
...
+						}
+					},
+				)
 				genesis.GenesisTime = GenesisTime
 				return genesis
 			}).Build()
 			ctx := tApp.InitChain()
 
+			currencyPair, _ := slinky.MarketPairToCurrencyPair(testMarketParam.Param.Pair)
+			market, _ := tApp.App.MarketMapKeeper.GetMarket(ctx, currencyPair.String())
```
Comment (vincentwschau) [inline protocol/testing/e2e/gov/add_new_market_test.go:322]: nit: Check that err is `nil`?

File: protocol/testing/e2e/gov/prices_test.go [hunk 2]
```diff
@@ -140,6 +142,34 @@ func TestUpdateMarketParam(t *testing.T) {
 						genesisState.Params.VotingPeriod = &testapp.TestVotingPeriod
 					},
 				)
+				// Initialize marketmap module with genesis market.
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *marketmaptypes.GenesisState) {
+						markets := make(map[string]marketmaptypes.Market)
+						cp, _ := slinky.MarketPairToCurrencyPair(GENESIS_MARKET_PARAM.Pair)
+						ticker := marketmaptypes.Ticker{
```
Comment (vincentwschau) [inline protocol/testing/e2e/gov/prices_test.go:151]: nit: Add a constant to `testutil/constants` for this market map market?

File: protocol/x/perpetuals/lib/lib_test.go [hunk 3]
```diff
@@ -421,36 +421,16 @@ func TestGetMarginRequirementsInQuoteQuantums_2(t *testing.T) {
 		bigExpectedInitialMargin     *big.Int
 		bigExpectedMaintenanceMargin *big.Int
 	}{
-		"InitialMargin 2 BIPs, MaintenanceMargin 1 BIP, positive exponent, atomic resolution 8": {
```
Comment (vincentwschau) [inline protocol/x/perpetuals/lib/lib_test.go:424]: nit: Add TODO for revisiting adding back positive exponents once supported by x/marketmap?

File: protocol/x/prices/keeper/market.go [hunk 4]
```diff
@@ -57,6 +57,13 @@ func (k Keeper) CreateMarket(
 		)
 	}
 	currencyPairStr := currencyPair.String()
+	_, err = k.MarketMapKeeper.GetMarket(ctx, currencyPairStr)
```
Comment (vincentwschau) [inline protocol/x/prices/keeper/market.go:60]: Are there tests for this new logic?

File: protocol/testing/e2e/gov/add_new_market_test.go (hunk 1 above)
Comment (chenyaoy) [reply protocol/testing/e2e/gov/add_new_market_test.go]: done

File: protocol/x/perpetuals/lib/lib_test.go (hunk 3 above)
Comment (chenyaoy) [reply protocol/x/perpetuals/lib/lib_test.go:424]: done

File: protocol/x/prices/simulation/genesis.go [hunk 5]
```diff
@@ -80,14 +88,25 @@ func RandomizedGenState(simState *module.SimulationState) {
...
+		exchangeJson := fmt.Sprintf(exchangeJsonTemplate, marketName)
+
 		marketParams[i] = types.MarketParam{
 			Id:                uint32(i),
 			Pair:              marketName,
 			Exponent:          int32(marketExponent),
 			MinExchanges:      uint32(minExchanges),
 			MinPriceChangePpm: uint32(simtypes.RandIntBetween(r, 1, int(lib.MaxPriceChangePpm))),
 			// The simulation tests don't run the daemon currently so we pass in empty exchange config.
-			ExchangeConfigJson: "{}",
```
[the remark this answers is not in the document]
Comment (chenyaoy) [reply protocol/x/prices/simulation/genesis.go:90]: done

File: protocol/x/prices/keeper/market.go (hunk 4 above)
Comment (chenyaoy) [reply protocol/x/prices/keeper/market.go:60]: added an explicit check to market_test.go that error is not nil when creating new market before it exists in market map

File: protocol/testing/e2e/gov/prices_test.go (hunk 2 above)
Comment (chenyaoy) [reply protocol/testing/e2e/gov/prices_test.go:151]: This is derived from the GENESIS_MARKET_PARAM var defined in this file, so I think it makes sense to leave it like this

File: .github/workflows/indexer-build-and-push-dev-staging.yml [hunk 6]
```diff
@@ -6,6 +6,7 @@ on: # yamllint disable-line rule:truthy
       - main
       - 'release/indexer/v[0-9]+.[0-9]+.x'  # e.g. release/indexer/v0.1.x
       - 'release/indexer/v[0-9]+.x'  # e.g. release/indexer/v1.x
+      - 'chenyao/tra-508'
```
Comment (vincentwschau) [inline .github/workflows/indexer-build-and-push-dev-staging.yml:9]: nit: Remove before merging.

File: .github/workflows/protocol-build-and-push.yml [hunk 7]
```diff
@@ -6,6 +6,7 @@ on:  # yamllint disable-line rule:truthy
       - main
       - 'release/protocol/v[0-9]+.[0-9]+.x'  # e.g. release/protocol/v0.1.x
       - 'release/protocol/v[0-9]+.x'  # e.g. release/protocol/v1.x
+      - 'chenyao/tra-508'
```
Comment (vincentwschau) [inline .github/workflows/protocol-build-and-push.yml:9]: nit: Remove before merging.

File: protocol/x/prices/keeper/market_test.go [hunk 8]
```diff
@@ -28,6 +28,18 @@ func TestCreateMarket(t *testing.T) {
...
+	_, err := keeper.CreateMarket(
+		ctx,
+		testMarketParams,
+		types.MarketPrice{
+			Id:       0,
+			Exponent: int32(-6),
+			Price:    constants.FiveBillion,
+		},
+	)
+	require.Error(t, err)
```
Comment (vincentwschau) [inline protocol/x/prices/keeper/market_test.go:41]: nit: Check the error against the expected error type.

**Review by vincentwschau — APPROVED** [review]

**Pull request merged**